src/Dependencies/Collections/SegmentedHashSet`1.cs

```diff
@@ -39,70 +49,60 @@ public class HashSet<T> : ICollection<T>, ISet<T>, IReadOnlyCollection<T>, IRead
 
         private int[]? _buckets;
         private Entry[]? _entries;
-#if TARGET_64BIT
         private ulong _fastModMultiplier;
-#endif
         private int _count;
         private int _freeList;
         private int _freeCount;
         private int _version;
-        private IEqualityComparer<T>? _comparer;
-        private SerializationInfo? _siInfo; // temporary variable needed during deserialization
+#if NETCOREAPP
+        private readonly IEqualityComparer<T>? _comparer;
+#else
+        /// <summary>
+        /// <see cref="EqualityComparer{T}.Default"/> doesn't devirtualize on .NET Framework, so we always ensure
+        /// <see cref="_comparer"/> is initialized to a non-<see langword="null"/> value.
+        /// </summary>
+        private readonly IEqualityComparer<T> _comparer;
+#endif
 
         #region Constructors
 
-        public HashSet() : this((IEqualityComparer<T>?)null) { }
+        public SegmentedHashSet() : this((IEqualityComparer<T>?)null) { }
 
-        public HashSet(IEqualityComparer<T>? comparer)
+        public SegmentedHashSet(IEqualityComparer<T>? comparer)
         {
             if (comparer != null && comparer != EqualityComparer<T>.Default) // first check for null to avoid forcing default comparer instantiation unnecessarily
             {
                 _comparer = comparer;
             }
 
-            // Special-case EqualityComparer<string>.Default, StringComparer.Ordinal, and StringComparer.OrdinalIgnoreCase.
-            // We use a non-randomized comparer for improved perf, falling back to a randomized comparer if the
-            // hash buckets become unbalanced.
-
-            if (typeof(T) == typeof(string))
-            {
-                if (_comparer is null)
-                {
-                    _comparer = (IEqualityComparer<T>)NonRandomizedStringEqualityComparer.WrappedAroundDefaultComparer;
-                }
-                else if (ReferenceEquals(_comparer, StringComparer.Ordinal))
-                {
-                    _comparer = (IEqualityComparer<T>)NonRandomizedStringEqualityComparer.WrappedAroundStringComparerOrdinal;
-                }
-                else if (ReferenceEquals(_comparer, StringComparer.OrdinalIgnoreCase))
-                {
-                    _comparer = (IEqualityComparer<T>)NonRandomizedStringEqualityComparer.WrappedAroundStringComparerOrdinalIgnoreCase;
```

Comment:
Why were these special cases eliminated?

Reply:
They require porting a huge amount of platform-specific culture code. The same change was adopted for `SegmentedDictionary<TKey, TValue>`.